api/src/main/java/org/reactivestreamsio/tcp/TcpConnection.java

```diff
@@ -0,0 +1,172 @@
+ *
+ * <h2>Flush</h2>
+ *
+ * In order to optimize the number of system calls to write data on the underneath socket, this connection breaks the
+ * writing process into two parts, viz., write and flush. A flush call writes all previous writes on this connection on
+ * the underneath socket.
+ *
+ * <h2>Write errors</h2>
+ *
+ * Every write call returns a {@link Publisher} that represents the acknowledgment of the write action. If the write
+ * failed, then the subscriber's {@link Subscriber#onError(Throwable)} method will be called and if it succeeds, then
+ * subscriber's {@link Subscriber#onComplete()} method will be called
+ *
+ * <h2>Flush errors</h2>
+ *
+ * Since, a {@link #flush()} is a batch operation encompassing all previous unflushed writes, it will represent the
+ * result of all those writes. It will fail if any one of the writes fail and it will not be possible to correlate which
+ * write caused the flush to fail. In order to get granular result per write, one has to subscribe to the
+ * {@link Publisher} of that particular write.
+ *
+ * <h2>Close</h2>
+ *
+ * Since, this connection only allows a single subscription, cancelling that subscription
+ * (via {@link Subscription#cancel()}) closes the connection.
+ * For cases which ignores content (via {@link #ignoreContent()}), an explicit {@link #close()} is required to be
+ * invoked for closing this connection.
+ *
+ * @param <R> Type of objects read from this connection.
+ * @param <W> Type of objects written to this connection.
+ */
+public interface TcpConnection<R, W> extends Publisher<R> {
+
+    /**
+     * Subscribes to the content stream of this connection.
+     *
+     * <h2>Multiple subscriptions</h2>
+     * This connection only allows a single subscription to the content stream. Any subsequent subscription will result
+     * in invocation of {@link Subscriber#onError(Throwable)} on that subscriber.
+     *
+     * @param subscriber Subscriber for the content.
+     */
+    @Override
+    void subscribe(Subscriber<? super R> subscriber);
+
+    /**
+     * Ignores all content of this connection. This is equivalent to calling {@link #subscribe(Subscriber)} with a
+     * subscriber that discards all received items.
+     */
+    void ignoreContent();
+
+    /**
+     * On subscription of the returned {@link Publisher}, writes the passed {@code msg} on this connection.
+     * Every subscription will write the same {@code msg} on the connection.
+     *
+     * <b>This does not flush the write.</b> Call {@link #flush()} to flush this write.
+     * @param msg Message to write.
+     *
+     * @return A {@link Publisher} representing the result of this write. One has to subscribe to this {@link Publisher}
+     * to trigger the write. Each subscription will trigger a write of the same message. The subscriber to this
+     * {@link Publisher} will not complete unless {@link #flush()} is called on this connection.
+     */
+    Publisher<Void> write(W msg);
```

Comment:
>  In other words, it's acting as a Future.

Yes but not quiet like a Future! Futures are always hot and can only be used once. `Publisher` here and `Observable` in RxJava are cold (in most cases) i.e. one needs to subscribe to them to start execution. Also, one can subscribe to them multiple times. In this case, multiple subscriptions would trigger multiple write+flush.

> Is it standard practice in Java RS-based interfaces to use a 0-item or 1-item Publisher as a Future? Or just in standard interfaces which need to be usable by any RS implementation? It just looks overly complex, and possibly more expensive than a Future.

I generally prefer using a single async abstraction and hence use `Observable` everywhere. In case of interfacing with netty, I bridged a `Subscriber` and `Future` from netty.
It takes a bit of getting used to treating a scalar response as a stream of 1 but once you wrap your head around it, it has a benefit of consistency. FWIW, RxJava has this issue https://github.com/ReactiveX/RxJava/issues/1594 to introduce Future/Task in RxJava.